[Bug]: Contains and EndsWith should not be mandatory, but the TCK makes it mandatory.

### Specification Version

1.0.0

### Bug report

By specification:

> An implementation of Query by Method Name backed by a document or graph database is not required to support Contains, EndsWith, StartsWith, Like, IgnoreCase, In, or Null. A repository method must raise java.lang.UnsupportedOperationException or a more specific subclass of the exception if the database does not provide the requested functionality.

Ref: https://github.com/jakartaee/data/blob/main/spec/src/main/asciidoc/method-query.asciidoc#27-query-by-method-name-conditions


However, when I ran at the TCK, it shows it as an error:

```
[ERROR]   MyEntityTests>EntityTests.testContainsInString:419 » UnsupportedOperation Contains is not supported in Eclipse JNoSQL method query
[ERROR]   MyEntityTests>EntityTests.testFindFirst3:617 » UnsupportedOperation EndsWith is not supported in Eclipse JNoSQL method query
```



### Additional information

_No response_

--- a/tck-dist/pom.xml
+++ b/tck-dist/pom.xml
@@ -92,7 +92,7 @@
     <artifactId>maven-dependency-plugin</artifactId>
     <version>3.6.1</version>
     <executions>
-	<!-- Load TCK dependency location as variable -->
+    <!-- Load TCK dependency location as variable -->
      <execution>
       <phase>initialize</phase>
       <goals>

--- a/tck-dist/src/main/pom.xml
+++ b/tck-dist/src/main/pom.xml
@@ -30,8 +30,8 @@
  <version>1.0-SNAPSHOT</version>
   
   <modules>
-	  <module>starter/ee-pom.xml</module>
-	  <module>starter/se-pom.xml</module>
+      <module>starter/ee-pom.xml</module>
+      <module>starter/se-pom.xml</module>
   </modules>
  
  </project>

--- a/tck/pom.xml
+++ b/tck/pom.xml
@@ -104,7 +104,7 @@
       <version>${jakarta.transaction.version}</version>
       <scope>provided</scope>
     </dependency>
-	
+    
     <dependency>
       <groupId>jakarta.validation</groupId>
       <artifactId>jakarta.validation-api</artifactId>

--- a/tck/src/main/java/ee/jakarta/tck/data/framework/utilities/DatabaseType.java
+++ b/tck/src/main/java/ee/jakarta/tck/data/framework/utilities/DatabaseType.java
@@ -21,7 +21,7 @@ import java.util.Arrays;
  * This enum represents the configured DatabaseType based on the {@link TestProperty} databaseType
  */
 public enum DatabaseType {
-    UNKNOWN(Integer.MAX_VALUE), //No database type was configured
+    OTHER(Integer.MAX_VALUE), //No database type was configured
     RELATIONAL(100),
     GRAPH(50),
     DOCUMENT(40),
@@ -36,7 +36,7 @@ public enum DatabaseType {
     }
     
     public static DatabaseType valueOfIgnoreCase(String value) {
-        return Arrays.stream(DatabaseType.values()).filter(type -> type.name().equalsIgnoreCase(value)).findAny().orElse(DatabaseType.UNKNOWN);
+        return Arrays.stream(DatabaseType.values()).filter(type -> type.name().equalsIgnoreCase(value)).findAny().orElse(DatabaseType.OTHER);
     }
     
     public boolean isKeywordSupportAtOrBelow(DatabaseType benchmark) {

--- a/tck/src/main/java/ee/jakarta/tck/data/framework/utilities/DatabaseType.java
+++ b/tck/src/main/java/ee/jakarta/tck/data/framework/utilities/DatabaseType.java
@@ -1,0 +1,34 @@
+/*
+ * Copyright (c) 2024 Contributors to the Eclipse Foundation
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0, which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception, which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ */
+package ee.jakarta.tck.data.framework.utilities;
+
+import java.util.Arrays;
+
+/**
+ * This enum represents the configured DatabaseType based on the {@link TestProperty} databaseType
+ */
+public enum DatabaseType {
+    UNKNOWN, //No database type was configured
+    RELATIONAL,
+    COLUMN,
+    DOCUMENT,
+    GRAPH,
+    KEY_VALUE;
+    
+    public static DatabaseType valueOfIgnoreCase(String value) {
+        return Arrays.stream(DatabaseType.values()).filter(type -> type.name().equalsIgnoreCase(value)).findAny().orElse(DatabaseType.UNKNOWN);
+    }
+}

--- a/tck/src/main/java/ee/jakarta/tck/data/framework/utilities/DatabaseType.java
+++ b/tck/src/main/java/ee/jakarta/tck/data/framework/utilities/DatabaseType.java
@@ -21,14 +21,25 @@ import java.util.Arrays;
  * This enum represents the configured DatabaseType based on the {@link TestProperty} databaseType
  */
 public enum DatabaseType {
-    UNKNOWN, //No database type was configured
-    RELATIONAL,
-    COLUMN,
-    DOCUMENT,
-    GRAPH,
-    KEY_VALUE;
+    UNKNOWN(Integer.MAX_VALUE), //No database type was configured
+    RELATIONAL(100),
+    GRAPH(50),
+    DOCUMENT(40),
+    TIME_SERIES(30),
+    COLUMN(20),
+    KEY_VALUE(10);
+    
+    private int flexibility;
+    
+    private DatabaseType(int flexibility) {
+        this.flexibility = flexibility;
+    }
     
     public static DatabaseType valueOfIgnoreCase(String value) {
         return Arrays.stream(DatabaseType.values()).filter(type -> type.name().equalsIgnoreCase(value)).findAny().orElse(DatabaseType.UNKNOWN);
+    }
+    
+    public boolean isKeywordSupportAtOrBelow(DatabaseType benchmark) {
+        return this.flexibility <= benchmark.flexibility;
     }
 }

--- a/tck/src/main/java/ee/jakarta/tck/data/framework/utilities/TestProperty.java
+++ b/tck/src/main/java/ee/jakarta/tck/data/framework/utilities/TestProperty.java
@@ -31,33 +31,49 @@ import ee.jakarta.tck.data.framework.arquillian.extensions.TCKFrameworkAppender;
  */
 public enum TestProperty {    
     //Java properties that should always be set by the JVM
-    javaHome    (true,  "java.home",                  "Path to the java executable used to create the current JVM"),
-    javaSpecVer (true,  "java.specification.version", "Specification version of the java executable"),
-    javaTempDir (true,  "java.io.tmpdir",             "The path to a temporary directory where a copy of the signature file will be created"),
-    javaVer     (true,  "java.version",               "Full version of the java executable"),
+    javaHome    (true,  "java.home",
+            "Path to the java executable used to create the current JVM"),
+    javaSpecVer (true,  "java.specification.version",
+            "Specification version of the java executable"),
+    javaTempDir (true,  "java.io.tmpdir",
+            "The path to a temporary directory where a copy of the signature file will be created"),
+    javaVer     (true,  "java.version",
+            "Full version of the java executable"),
     
     //TCK specific properties
-    skipDeployment (false,  "jakarta.tck.skip.deployment",  "If true, run in SE mode and do not use Arquillian deployment, if false run in EE mode and use Arquillian deployments. "
+    skipDeployment (false,  "jakarta.tck.skip.deployment",
+            "If true, run in SE mode and do not use Arquillian deployment, if false run in EE mode and use Arquillian deployments. "
             + "Default: false", "false"),
-    pollFrequency  (false, "jakarta.tck.poll.frequency",    "Time in seconds between polls of the repository to verify read-only data was successfully written. "
+    pollFrequency  (false, "jakarta.tck.poll.frequency",
+            "Time in seconds between polls of the repository to verify read-only data was successfully written. "
             + "Default: 1 second", "1"),
-    pollTimeout    (false, "jakarta.tck.poll.timeout",      "Time in seconds when we will stop polling to verify read-only data was successfully written. "
+    pollTimeout    (false, "jakarta.tck.poll.timeout",
+            "Time in seconds when we will stop polling to verify read-only data was successfully written. "
             + "Default: 60 seconds", "60"),
-    delay          (false, "jakarta.tck.consistency.delay", "Time in seconds after verifying read-only data was successfully written to respository "
-            + "for repository to have consistency. Default: none", ""),
+    delay          (false, "jakarta.tck.consistency.delay",
+            "Time in seconds after verifying read-only data was successfully written to respository "
+            + "for repository to have consistency. "
+            + "Default: none"),
+    databaseType   (false, "jakarta.tck.database.type",
+            "The type of database being used. Valid values are " + DatabaseType.values() 
+            + " (case insenstive). The database type is used to make assertions based on the underlying database. "
+            + "Default: UNKNOWN", "UNKNOWN"),
+    databaseName   (false, "jakarta.tck.database.name",
+            "The name of database being used. The database name is used to make assertions based on the underlying database. "
+            + "Default: none"),
     
     //Signature testing properties
     signatureClasspath (false,  "signature.sigTestClasspath", "The path to the Jakarta Data API JAR used by your implementation. "
-            + "Required for standalone testing, but optional when testing on a Jakarta EE profile.", ""),
+            + "Required for standalone testing, but optional when testing on a Jakarta EE profile. "
+            + "Default: none"),
     signatureImageDir  (true,   "jimage.dir",                 "The path to a directory that is readable and writable that "
-            + "the signature test will cache Java SE modules as classes");
+            + "the signature test will cache Java SE modules as classes. "
+            + "Default: none");
     
-    private static final Logger log = Logger.getLogger(TestProperty.class.getCanonicalName());
-
     private boolean required;
     private String key;
-    private String description;
-    private String defaultValue;    
+    private String value;
+    private String description;   
     
     // CONSTRUCTORS
     private TestProperty(boolean required, String key, String description) {
@@ -68,7 +84,7 @@ public enum TestProperty {
         this.required = required;
         this.key = key;
         this.description = description;
-        this.defaultValue = defaultValue;
+        this.value = getValue(defaultValue);
     }
     
     // GETTERS
@@ -90,7 +106,6 @@ public enum TestProperty {
     }
     
     public boolean isSet() {
-        String value = getValue(false);
         if(value == null)
             return false;
         if(value.isBlank() || value.isEmpty()) {
@@ -101,11 +116,19 @@ public enum TestProperty {
     
     // CONVERTERS
     public long getLong() throws IllegalStateException, NumberFormatException {
-        return Long.parseLong(getValue());
+        return Long.parseLong(value);
     }
     
     public int getInt() throws IllegalStateException, NumberFormatException {
-        return Integer.parseInt(getValue());
+        return Integer.parseInt(value);
+    }
+    
+    public boolean getBoolean() {
+        return Boolean.parseBoolean(value);
+    }
+    
+    public DatabaseType getDatabaseType() {
+        return DatabaseType.valueOfIgnoreCase(value);
     }
     
     /**
@@ -114,35 +137,41 @@ public enum TestProperty {
      * @return the property value
      * @throws IllegalStateException if required and no property was found
      */
-    public String getValue() throws IllegalStateException {
-        return getValue(required);
-    }
-    
-    private String getValue(boolean verify) throws IllegalStateException {
-        String value = null;
-        log.fine("Searching for property: " + key);
-        
-        // Client: get property from system 
-        if(value == null) {
-            value = System.getProperty(key);
-            log.fine("Value from system: " + value);
-        }
-        
-        //Container: get property from properties file
-        if(value == null) {
-            value = TestPropertyHandler.loadProperties().getProperty(key);
-            log.fine("Value from resource file: " + value);
-        }
-        
-        //Default: get default property
-        if(value == null) {
-            value = defaultValue;
-            log.fine("Defaulting to value: " + value);
-        }
-        
-        if(verify && value == null)
+    public String getValue() {
+        if(required && value == null)
             throw new IllegalStateException("Could not obtain a value for system property: " + key);
         
         return value;
+    }
+    
+    private String getValue(String defaultVal) throws IllegalStateException {
+        final Logger log = Logger.getLogger(TestProperty.class.getCanonicalName());
+        
+        String valueLocal = null;
+        log.fine("Searching for property: " + key);
+        
+        // Client: get property from system 
+        if(valueLocal == null) {
+            valueLocal = System.getProperty(key);
+            log.fine("Value from system: " + valueLocal);
+        }
+        
+        //Container: get property from properties file
+        if(valueLocal == null) {
+            valueLocal = TestPropertyHandler.loadProperties().getProperty(key);
+            log.fine("Value from resource file: " + valueLocal);
+        }
+        
+        //Default: get default property
+        if(valueLocal == null) {
+            valueLocal = defaultVal;
+            log.fine("Value set to default: " + valueLocal);
+        }
+        
+        if (valueLocal == null) {
+            log.fine("Property was not set, value: " + null);
+        }
+        
+        return valueLocal;
     }
 }

--- a/tck/src/main/java/ee/jakarta/tck/data/framework/utilities/TestProperty.java
+++ b/tck/src/main/java/ee/jakarta/tck/data/framework/utilities/TestProperty.java
@@ -56,9 +56,9 @@ public enum TestProperty {
             + "for repository to have consistency. "
             + "Default: none"),
     databaseType   (false, "jakarta.tck.database.type",
-            "The type of database being used. Valid values are " + Arrays.asList(DatabaseType.values()).toString()
-            + " (case insensitive). The database type is used to make assertions based on the underlying database. "
-            + "Default: UNKNOWN", "UNKNOWN"),
+            "The type of database being used. Valid values are " + Arrays.asList(DatabaseType.values()).toString() + " (case insensitive). "
+            + "The database type is used to make assertions based on the keywords supported by the underlying database. "
+            + "Default: OTHER", "OTHER"),
     databaseName   (false, "jakarta.tck.database.name",
             "The name of database being used. The database name is used to make assertions based on the underlying database. "
             + "Default: none"),

--- a/tck/src/main/java/ee/jakarta/tck/data/framework/utilities/TestProperty.java
+++ b/tck/src/main/java/ee/jakarta/tck/data/framework/utilities/TestProperty.java
@@ -15,6 +15,7 @@
  */
 package ee.jakarta.tck.data.framework.utilities;
 
+import java.util.Arrays;
 import java.util.logging.Logger;
 
 import ee.jakarta.tck.data.framework.arquillian.extensions.TCKFrameworkAppender;
@@ -55,8 +56,8 @@ public enum TestProperty {
             + "for repository to have consistency. "
             + "Default: none"),
     databaseType   (false, "jakarta.tck.database.type",
-            "The type of database being used. Valid values are " + DatabaseType.values() 
-            + " (case insenstive). The database type is used to make assertions based on the underlying database. "
+            "The type of database being used. Valid values are " + Arrays.asList(DatabaseType.values()).toString()
+            + " (case insensitive). The database type is used to make assertions based on the underlying database. "
             + "Default: UNKNOWN", "UNKNOWN"),
     databaseName   (false, "jakarta.tck.database.name",
             "The name of database being used. The database name is used to make assertions based on the underlying database. "

--- a/tck/src/main/java/ee/jakarta/tck/data/framework/utilities/TestPropertyHandler.java
+++ b/tck/src/main/java/ee/jakarta/tck/data/framework/utilities/TestPropertyHandler.java
@@ -83,7 +83,8 @@ public class TestPropertyHandler {
             if(prop.getKey().startsWith("java.")) {
                 continue;
             }
-            filteredProps.put(prop.getKey(), prop.getValue());
+            if(prop.isSet())
+                filteredProps.put(prop.getKey(), prop.getValue());
         }
         
         try {

--- a/tck/src/main/java/ee/jakarta/tck/data/standalone/entity/EntityTests.java
+++ b/tck/src/main/java/ee/jakarta/tck/data/standalone/entity/EntityTests.java
@@ -57,6 +57,8 @@ import ee.jakarta.tck.data.framework.read.only.NaturalNumbers;
 import ee.jakarta.tck.data.framework.read.only.NaturalNumbersPopulator;
 import ee.jakarta.tck.data.framework.read.only.PositiveIntegers;
 import ee.jakarta.tck.data.framework.read.only.NaturalNumber.NumberType;
+import ee.jakarta.tck.data.framework.utilities.DatabaseType;
+import ee.jakarta.tck.data.framework.utilities.TestProperty;
 import ee.jakarta.tck.data.framework.utilities.TestPropertyUtility;
 import jakarta.data.Limit;
 import jakarta.data.Order;
@@ -111,6 +113,8 @@ public class EntityTests {
         assertNotNull(characters);
         AsciiCharactersPopulator.get().populate(characters);
     }
+    
+    private DatabaseType type = TestProperty.databaseType.getDatabaseType();
 
     @Assertion(id = "136", strategy = "Ensures that the prepopulation step for readonly entities was successful")
     public void ensureNaturalNumberPrepopulation() {
@@ -416,7 +420,16 @@ public class EntityTests {
 
     @Assertion(id = "133", strategy = "Use a repository method with Contains to query for a substring of a String attribute.")
     public void testContainsInString() {
-        Collection<AsciiCharacter> found = characters.findByHexadecimalContainsAndIsControlNot("4", true);
+        Collection<AsciiCharacter> found;
+        try {
+            found = characters.findByHexadecimalContainsAndIsControlNot("4", true);    
+        } catch (UnsupportedOperationException e) {
+            if(type == DatabaseType.DOCUMENT || type == DatabaseType.GRAPH) {
+                return; //passed
+            }
+            throw e;
+        }
+        
 
         assertEquals(List.of("24", "34",
                              "40", "41", "42", "43",
@@ -614,7 +627,17 @@ public class EntityTests {
     @Assertion(id = "133",
                strategy = "Use a repository method with findFirst3By that returns the first 3 results.")
     public void testFindFirst3() {
-        AsciiCharacter[] found = characters.findFirst3ByNumericValueGreaterThanEqualAndHexadecimalEndsWith(40, "4", Sort.asc("numericValue"));
+        AsciiCharacter[] found;
+        
+        try {
+        found = characters.findFirst3ByNumericValueGreaterThanEqualAndHexadecimalEndsWith(40, "4", Sort.asc("numericValue"));
+        } catch (UnsupportedOperationException e) {
+            if(type == DatabaseType.DOCUMENT || type == DatabaseType.GRAPH) {
+                return; //passed
+            }
+            throw e;
+        }
+        
         assertEquals(3, found.length);
         assertEquals('4', found[0].getThisCharacter());
         assertEquals('D', found[1].getThisCharacter());

--- a/tck/src/main/java/ee/jakarta/tck/data/standalone/entity/EntityTests.java
+++ b/tck/src/main/java/ee/jakarta/tck/data/standalone/entity/EntityTests.java
@@ -424,7 +424,7 @@ public class EntityTests {
         try {
             found = characters.findByHexadecimalContainsAndIsControlNot("4", true);    
         } catch (UnsupportedOperationException e) {
-            if(type == DatabaseType.DOCUMENT || type == DatabaseType.GRAPH) {
+            if(type.isKeywordSupportAtOrBelow(DatabaseType.DOCUMENT)) {
                 return; //passed
             }
             throw e;
@@ -632,7 +632,7 @@ public class EntityTests {
         try {
         found = characters.findFirst3ByNumericValueGreaterThanEqualAndHexadecimalEndsWith(40, "4", Sort.asc("numericValue"));
         } catch (UnsupportedOperationException e) {
-            if(type == DatabaseType.DOCUMENT || type == DatabaseType.GRAPH) {
+            if(type.isKeywordSupportAtOrBelow(DatabaseType.DOCUMENT)) {
                 return; //passed
             }
             throw e;

--- a/tck/src/main/java/ee/jakarta/tck/data/standalone/signature/SignatureTests.java
+++ b/tck/src/main/java/ee/jakarta/tck/data/standalone/signature/SignatureTests.java
@@ -24,6 +24,7 @@ import ee.jakarta.tck.data.framework.junit.anno.Assertion;
 import ee.jakarta.tck.data.framework.junit.anno.Signature;
 import ee.jakarta.tck.data.framework.junit.anno.Standalone;
 import ee.jakarta.tck.data.framework.signature.DataSignatureTestRunner;
+import ee.jakarta.tck.data.framework.utilities.TestProperty;
 
 @Standalone
 @AnyEntity
@@ -36,7 +37,7 @@ public class SignatureTests {
 
     @Assertion(id = "26", strategy = "Uses the sigtest-maven-plugin to execute signature tests on a Standalone JVM or on a Jakarta EE Server")
     public void testSignatures() throws Exception {
-        DataSignatureTestRunner.assertProjectSetup(true);
+        DataSignatureTestRunner.assertProjectSetup(TestProperty.skipDeployment.getBoolean());
         DataSignatureTestRunner runner = new DataSignatureTestRunner();
         runner.signatureTest();
     }

